### PR TITLE
Johtaylo/dialog samples projects

### PR DESCRIPTION
**Pull request by johnataylor**
This is _just the projects_ for three samples. Not the full implementation.
- simple-prompt-bot
- multiturn-prompt-bot
- custom-dialog

### CODE REVIEW

File: csharp_dotnetcore/4.Simple-Prompt-Bot/BotBuilder.ruleset [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  </Rules>
+  <Rules AnalyzerId="Microsoft.CodeAnalysis.CSharp.Features" RuleNamespace="Microsoft.CodeAnalysis.CSharp.Features">
+    <Rule Id="IDE0003" Action="None" />
+  </Rules>
+  <Rules AnalyzerId="StyleCop.Analyzers" RuleNamespace="StyleCop.Analyzers">
+    <Rule Id="SA1101" Action="None" />   <!-- local calls prefixed with this -->
+    <Rule Id="SA1129" Action="None" />   <!-- don't use value type constructor-->
+    <Rule Id="SA1200" Action="None" />
+    <Rule Id="SA1214" Action="None" />   <!-- encapsulate field in property-->
+
```
Comment (JasonSowers) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/BotBuilder.ruleset]: extra line

File: csharp_dotnetcore/21.Custom-Dialogs/Controllers/ValuesController.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+
+namespace _21.Custom_Dialogs.Controllers
+{
+    [Route("api/[controller]")]
```
Comment (JasonSowers) [inline csharp_dotnetcore/21.Custom-Dialogs/Controllers/ValuesController.cs]: Is this file needed? No copyright header if it is.

File: csharp_dotnetcore/21.Custom-Dialogs/21.Custom-Dialogs.csproj [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.8" />
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.0.0.38762" />
```
Comment (JasonSowers) [inline csharp_dotnetcore/21.Custom-Dialogs/21.Custom-Dialogs.csproj]: stylecop not added

File: csharp_dotnetcore/21.Custom-Dialogs/MainBot.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+
+namespace Custom_Dialogs
+{
+    /// <summary>
+    /// This bot illustrates a simple prompt.
```
Comment (JasonSowers) [inline csharp_dotnetcore/21.Custom-Dialogs/MainBot.cs]: I would lean towards this comment should be more robust

File: csharp_dotnetcore/4.Simple-Prompt-Bot/4.Simple-Prompt-Bot.csproj [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Folder Include="wwwroot\" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.8" />
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.0.0.38762" />
```
Comment (JasonSowers) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/4.Simple-Prompt-Bot.csproj]: add stylecop

File: csharp_dotnetcore/4.Simple-Prompt-Bot/BotAccessors.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the MIT License.
+
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Dialogs;
+
+namespace Simple_Prompt_Bot
+{
+    public class BotAccessors
+    {
+        public IStatePropertyAccessor<DialogState> ConversationDialogState { get; set; }
```
Comment (JasonSowers) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/BotAccessors.cs]: Add xml comments

File: csharp_dotnetcore/4.Simple-Prompt-Bot/SimplePromptBot.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>>A <see cref="Task"/> representing the operation result of the Turn operation.</returns>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (turnContext == null)
+            {
+                throw new ArgumentNullException(nameof(turnContext));
+            }
+
+            // We are only interested in Message Activities.
```
Comment (JasonSowers) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/SimplePromptBot.cs]: Why are we only interested in the ActivityType of message?

File: csharp_dotnetcore/5.MultiTurn-Prompts-Bot/5.MultiTurn-Prompts-Bot.csproj [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.8" />
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.0.0.38762" />
```
Comment (JasonSowers) [inline csharp_dotnetcore/5.MultiTurn-Prompts-Bot/5.MultiTurn-Prompts-Bot.csproj]: add stylecop

File: csharp_dotnetcore/5.MultiTurn-Prompts-Bot/BotAccessors.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Dialogs;
+
+namespace MultiTurn_Prompts_Bot
+{
```
Comment (JasonSowers) [inline csharp_dotnetcore/5.MultiTurn-Prompts-Bot/BotAccessors.cs]: needs commenting

File: csharp_dotnetcore/5.MultiTurn-Prompts-Bot/Controllers/ValuesController.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+
+namespace _5.MultiTurn_Prompts_Bot.Controllers
```
Comment (JasonSowers) [inline csharp_dotnetcore/5.MultiTurn-Prompts-Bot/Controllers/ValuesController.cs]: Not sure this is needed

File: csharp_dotnetcore/4.Simple-Prompt-Bot/BotAccessors.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Dialogs;
+
+namespace Simple_Prompt_Bot
+{
+    public class BotAccessors
```
Comment (daveta) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/BotAccessors.cs]: Might steal what we did in sample 2, we created a constructor that took the State object so we could signal that the State object should be accessible from the bot to enable explicit Save operations (we removed the BotStateSet).

File: csharp_dotnetcore/4.Simple-Prompt-Bot/SimplePromptBot.bot [hunk 12]
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "Simple-Prompt-Bot",
+    "description": "",
+    "secretKey": "",
+    "services": [
+        {
+            "appId": "",
+            "id": "http://localhost:1499/api/messages",
```
Comment (daveta) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/SimplePromptBot.bot]: >1499 [](start = 36, length = 4)

Think we're standardizing on port 3978

File: csharp_dotnetcore/4.Simple-Prompt-Bot/Startup.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        /// Initializes a new instance of the <see cref="Startup"/> class.
+        /// This method gets called by the runtime. Use this method to add services to the container.
+        /// For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940.
+        /// </summary>
+        /// <param name="env">Provides information about the <see cref="IHostingEnvironment"/> an application is running in.</param>
+        public Startup(IHostingEnvironment env)
+        {
+            var builder = new ConfigurationBuilder()
+                .SetBasePath(env.ContentRootPath)
+                .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
```
Comment (daveta) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/Startup.cs]: >AddJsonFile [](start = 17, length = 11)

Think we're cutting all ties to appsettings - we think this is possible.

File: csharp_dotnetcore/4.Simple-Prompt-Bot/Properties/launchSettings.json [hunk 14]
```diff
@@ -0,0 +1,27 @@
+{
+  "iisSettings": {
+    "windowsAuthentication": false,
+    "anonymousAuthentication": true,
+    "iisExpress": {
+      "applicationUrl": "http://localhost:1499/",
```
Comment (daveta) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/Properties/launchSettings.json]: >1499 [](start = 42, length = 4)

Might switch to 3978

File: csharp_dotnetcore/4.Simple-Prompt-Bot/Startup.cs [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+                // https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
+                // Uncomment this line to use Azure Blob Storage
+                // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
+                // Create and add conversation state.
+                var convoState = new ConversationState(dataStore);
+                options.State.Add(convoState);
+
+                // The BotStateSet middleware forces state storage to auto-save when the bot is complete processing the message.
+                // Note: Developers may choose not to add all the state providers to this middleware if save is not required.
+                var stateSet = new BotStateSet(options.State.ToArray());
```
Comment (daveta) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/Startup.cs]: >BotStateSet [](start = 35, length = 11)

Just given controversy, think we're cutting ties with this.
In sample 2, actually explicitly saved on the state object manually (before the response send).  Super conservative, but perhaps we should set this trend now :)

File: csharp_dotnetcore/4.Simple-Prompt-Bot/Startup.cs [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+                // is restarted, anything stored in memory will be gone.
+                IStorage dataStore = new MemoryStorage();
+
+                // For production bots use the Azure Blob or
+                // Azure CosmosDB storage providers, as seen below. To the Azure
+                // based storage providers, add the Microsoft.Bot.Builder.Azure
+                // Nuget package to your solution. That package is found at:
+                // https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
+                // Uncomment this line to use Azure Blob Storage
+                // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
```
Comment (daveta) [inline csharp_dotnetcore/4.Simple-Prompt-Bot/Startup.cs]: Might nuke :)

File: csharp_dotnetcore/5.MultiTurn-Prompts-Bot/BotAccessors.cs [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+    public class BotAccessors
```
Comment (daveta) [inline csharp_dotnetcore/5.MultiTurn-Prompts-Bot/BotAccessors.cs]: >BotAccessors [](start = 17, length = 12)

Same deal with constructor-giving-the-state object (see example 2 in core), and explicitly saving on state object before sending response so we set a conservative example.

File: csharp_dotnetcore/4.Simple-Prompt-Bot/Startup.cs (hunk 16 above)
Comment (johnataylor) [reply csharp_dotnetcore/4.Simple-Prompt-Bot/Startup.cs]: yes we did

File: csharp_dotnetcore/4.Simple-Prompt-Bot/Properties/launchSettings.json (hunk 14 above)
Comment (johnataylor) [reply csharp_dotnetcore/4.Simple-Prompt-Bot/Properties/launchSettings.json]: we left this for now

File: csharp_dotnetcore/4.Simple-Prompt-Bot/SimplePromptBot.cs (hunk 7 above)
Comment (johnataylor) [reply csharp_dotnetcore/4.Simple-Prompt-Bot/SimplePromptBot.cs]: here yes

File: csharp_dotnetcore/4.Simple-Prompt-Bot/4.Simple-Prompt-Bot.csproj (hunk 5 above)
Comment (johnataylor) [reply csharp_dotnetcore/4.Simple-Prompt-Bot/4.Simple-Prompt-Bot.csproj]: done

File: csharp_dotnetcore/21.Custom-Dialogs/MainBot.cs (hunk 4 above)
Comment (johnataylor) [reply csharp_dotnetcore/21.Custom-Dialogs/MainBot.cs]: yes done
